pravega/pravega
# Issue 1665: TCP based `ruok` health check and segmentstore implementation

**Change log description**
This PR implements a TCP based health check. 

This is based on zookeeper implementation of the same functionality. Please see here:
https://zookeeper.apache.org/doc/r3.4.10/zookeeperAdmin.html#sc_zkCommands

The health reporter consists of two parts, a health processor and a health reporter base class.

Different components can implement `HealthReporter` interface. Each independent component is represented by an unique id. This id can be hierarchical (separated with `/`). Each component can register its  own commands to the `HealthRequestProcessor` interface.

The `HealthRequestProcessor` responds to TCP connection (e.g. `nc` or `telnet`).
The cmds can be global or targetted towards a known reporter id.

e.g. A global command:

```
usxxkandham2:pravega kandha$ echo "tgts" | nc localhost 7070
segmentstore/inmemlog/3 { ruok  } 
segmentstore/inmemlog/2 { ruok  } 
segmentstore/cache/Container_0 { ruok  } 
segmentstore/cache/Container_1 { ruok  } 
segmentstore/cache/Container_2 { ruok  } 
segmentstore/cache/Container_3 { ruok  } 
segmentstore/inmemlog/1 { ruok  } 
segmentstore/inmemlog/0 { ruok  } 
```
or
```
usxxkandham2:pravega kandha$ echo "ruok" | nc localhost 7070
imok
```

A health check targetted towards a component:
```
usxxkandham2:pravega kandha$ echo "ruok segmentstore/inmemlog/1" | nc localhost 7070
imok
```

**Purpose of the change**
Implement health reporter components and segment store implementations.

**What the code does**
This PR fixes #1665.

**How to verify it**
1. Unit tests for health reporter pass.
2. Run `standalone`  and use `nc` / `telnet` to  check health of different components.
